doc/specs/#888 - Com Api.md

```diff
@@ -0,0 +1,812 @@
+        // Creation of the GetCompositeAppCatalogOptions has to use CoCreateInstance rather than normal winrt initialization since it's created by an out of proc com server.
+        GetCompositeAppCatalogOptions getCompositeAppCatalogOptions = CreateGetCompositeAppCatalogOptions();
+        getCompositeAppCatalogOptions.Catalogs().Append(windowsCatalog);
+        getCompositeAppCatalogOptions.Catalogs().Append(installingCatalog);
+        // Specify that the search behavior is to only query for local packages.
+        // Since the local catalog that is open is InstallingPackages, this will only find a result if installAppId is currently installing.
+        getCompositeAppCatalogOptions.CompositeSearchBehavior(CompositeSearchBehavior::Installing);
+        AppCatalog compositeCatalog{ co_await appInstaller.GetCompositeAppCatalogAsync(getCompositeAppCatalogOptions) };
+        co_await compositeCatalog.OpenAsync();
+
+        FindPackagesOptions findPackagesOptions = CreateFindOptions();
+        PackageMatchFilter filter;
+        filter.IsAdditive = true;
+        filter.Field = PackageMatchField::Id;
+        filter.Type = MatchType::Exact;
+        filter.Value = installAppId;
+        findPackagesOptions.Filters().Append(filter);
+        FindPackagesResult findPackagesResult{ co_await compositeCatalog.FindPackagesAsync(findPackagesOptions) };
+        winrt::IVectorView<ResultMatch> matches = findPackagesResult.Matches();
+        Package package = matches.GetAt(0).Package();
+
+        if (package.IsInstalling())
+        {
+            m_installPackageOperation = package.InstallProgress();
+            UpdateUIForInstall(m_installPackageOperation, uiThread, button, progressBar, statusText); 
+            // Alternatively to Cancel the operation, use:
+            // m_installPackageOperation.Cancel();
+        }
+    }
+```
+
+## 3.6. Find sources
+
+```c++ (C++ish pseudocode)
+    IAsyncOperation<AppCatalog> FindDefaultSource()
+    {
+        AppInstaller appInstaller = CreateAppInstaller();
+
+        winrt::IVectorView<AppCatalog> catalogs{ co_await appInstaller.GetAppCatalogsAsync() };
+        for (AppCatalog catalog : catalogs)
+        {
+            if (catalog.Details().Origin == AppCatalogOrigin::Default)
+            {
+                co_await catalog.OpenAsync();
+                co_return catalog;
+            }
+        }
+        winrt::throw_hresult(E_UNEXPECTED);
+    }
+```
+
+## 3.7. Open a catalog by name
+
+Open a user-added catalog by name. There is no way to use the api to add a catalog, that must be done on the command line.
+
+```c++ (C++ish pseudocode)
+
+    IAsyncOperation<AppCatalog> FindSource(std::wstring packageSource)
+    {
+        AppInstaller appInstaller = CreateAppInstaller();
+        AppCatalog catalog{ co_await appInstaller.GetAppCatalogAsync(packageSource) };
+        co_await catalog.OpenAsync();

```

Comment:
It's a little strange that `OpenAsync` doesn't return anything, like "catalog not found" or "access denied" or something. Or are those errors coming out of `GetAppCatalogAsync` ?

Reply:
IAsyncAction has an ErrorCode property. OpenAsync definitely can have errors since it can hit the network, I just assumed the style was to use the IAsyncAction ErrorCode property rather than adding a new type that had it's own Errorcode when the internal errors are able to map to hresults (i.e. not like Http status codes or something).

Reply:
See https://www.osgwiki.com/wiki/WinRT_API_Design_Guidelines#Specifying_Asynchronous_Operation_Objects_in_IDL (non-MSFT folks, I hope to have these publicly visible soon) for how to do it.  The issue is that we want people to *not* wrap all their operations in try/catch.  When an operation fails and returns an hresult it generates an exception in most langauges, so you'd have to write this:

```c++
AppCatalog ac{nullptr};
winrt::hresult ecode;
try {
    ac = co_await foo.FindSource(sourceName);
} catch (winrt::hresult_exception const& ex) {
    ecode = ex.code();
}
if (!ac) {
    // do something with 'ecode'
} else {
    // catalog ready
}
```

Compare that to:

```c++
FindAppCatalogResult found = co_await thing.FindSource(sourceName);
if (found.Catalog) {
    // use found.Catalog
} else {
    switch (found.Status()) {
    case FindCatalogResultStatus::AccessDenied:
    case FindCatalogResultStatus::NotFound:
    case FindCatalogResultStatus::ServerError:
    case FindCatalogResultStatus::MalformedResponse:
        // do whatever with those
        return;
    case FindCatalogResultStatus::Unknown:
        // Send telemetry using found.ExtendedError();
        return;
    }
}
```